src/libopensc/sc-ossl-compat.h

```diff
@@ -114,208 +55,11 @@ extern "C" {
 #define EVP_PKEY_get_raw_public_key(p, pu, l)   (0)
 #endif
 
+/* OpenSSL 1.1.1 has FIPS_mode function */
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
-#define EC_POINT_get_affine_coordinates_GFp     EC_POINT_get_affine_coordinates
-#define EC_POINT_set_affine_coordinates_GFp     EC_POINT_set_affine_coordinates
-
-# ifndef FIPS_mode
 #define FIPS_mode()                             EVP_default_properties_is_fips_enabled(NULL)
-# endif
-
-/* As defined in openssl/include/openssl/evp.h */
-# ifndef EVP_PK_RSA
-#  define EVP_PK_RSA      0x0001
-#  define EVP_PK_DSA      0x0002
-#  define EVP_PK_DH       0x0004
-#  define EVP_PK_EC       0x0008
-#  define EVP_PKT_SIGN    0x0010
-#  define EVP_PKT_ENC     0x0020
-#  define EVP_PKT_EXCH    0x0040
-#  define EVP_PKS_RSA     0x0100
-#  define EVP_PKS_DSA     0x0200
-#  define EVP_PKS_EC      0x0400
```

Comment:
Some of those are still used in `p11test` and will be missing if OpenSSL is built with `no-deprecated`.

Reply:
There are a lot of warning about deprecated functions, as we have been targeting 1.1.1. All of those are in evp.h which is not included in some of the files that have changed.

```
doug@XUbuntu-20:/afs/anl.gov/appl/OpenSC-dev/build/opensc-git-my/amd64_linux26-ossl-3.0.1$ grep warning MA1.LOG 
../../../src/src/pkcs15init/pkcs15-westcos.c:249:6: warning: ‘EVP_PKEY_CTX_set_rsa_keygen_pubexp’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/pkcs15-pubkey.c:1858:5: warning: ‘EC_POINT_get_affine_coordinates_GFp’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/card-iasecc.c:3184:2: warning: ‘SHA1_Init’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/card-iasecc.c:3185:2: warning: ‘SHA1_Update’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/card-iasecc.c:3210:2: warning: ‘SHA1_Final’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/card-iasecc.c:3236:2: warning: ‘SHA256_Init’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/card-iasecc.c:3237:2: warning: ‘SHA256_Update’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/libopensc/card-iasecc.c:3262:2: warning: ‘SHA256_Final’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:232:2: warning: ‘ENGINE_by_id’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:244:3: warning: ‘ENGINE_by_id’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:247:4: warning: ‘ENGINE_by_id’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:249:3: warning: ‘ENGINE_ctrl_cmd_string’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:250:6: warning: ‘ENGINE_ctrl_cmd_string’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:251:4: warning: ‘ENGINE_free’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:257:3: warning: ‘ENGINE_set_default’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:258:3: warning: ‘ENGINE_free’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:232:2: warning: ‘ENGINE_by_id’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:244:3: warning: ‘ENGINE_by_id’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:247:4: warning: ‘ENGINE_by_id’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:249:3: warning: ‘ENGINE_ctrl_cmd_string’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:250:6: warning: ‘ENGINE_ctrl_cmd_string’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:251:4: warning: ‘ENGINE_free’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:257:3: warning: ‘ENGINE_set_default’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/pkcs11/openssl.c:258:3: warning: ‘ENGINE_free’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
../../../src/src/tools/pkcs15-init.c:1346:3: warning: ‘EVP_PKEY_cmp’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
```

Reply:
As pointed out, EVP_PK_RSA and EVP_PK_EC are deprecated in 3.0.0. They are only used in p11test in these routines:
 
./src/tests/p11test/p11test_case_common.c
./src/tests/p11test/p11test_case_readonly.c
./src/tests/p11test/p11test_case_pss_oaep.c

and only used to set "test_cert_t  type" in p11test_case_common.c:

 261         if (EVP_PKEY_base_id(evp) == EVP_PKEY_RSA) {
 262                 o->key = evp;
 263                 o->type = EVP_PK_RSA;
 264                 o->bits = EVP_PKEY_bits(evp);
 265
 266         } else if (EVP_PKEY_base_id(evp) == EVP_PKEY_EC) {
 267                 o->key = evp;
 268                 o->type = EVP_PK_EC;
 269                 o->bits = EVP_PKEY_bits(evp);
 270 
In other words the EVP_PK_* are only used internally to set a type and print error messages.

I would suggest either these "types" be maintained in p11tests or better still, 
EVP_PKEY_RSA and EVP_PKEY_EC be used instead and drop the use of these deprecated defines now.  

These are defined in evp.h in versions we are going to support: LibreSSL and OpenSSL 1.1.1 and 3.0.0:

# define EVP_PKEY_RSA    NID_rsaEncryption
# define EVP_PKEY_RSA2   NID_rsa
# define EVP_PKEY_RSA_PSS NID_rsassaPss
# define EVP_PKEY_EC	 NID_X9_62_id_ecPublicKey

Also note that EVP_PKEY_get_base_id() returns the type of pkey: EVP_PKEY_RSA
and EVP_PKEY_type() returns a specific NID. This might be useful for EVP_PKEY_RSA_PSS?

I would also suggest the evp.h be included specifically in the source code were needed.
Only p11test_case_wrap.c and p11test_case_pss_oaep.c include evp.h